src/snowflake/connector/cache.py

```diff
@@ -260,3 +319,218 @@ def _expiration(self, k: K) -> None:
         called from contexts where the lock is already held.
         """
         self.telemetry["expiration"] += 1
+
+    def _add_or_remove(self) -> None:
+        """This function gets called when an element is added, or removed.
+
+        Note that while this function does not interact with lock, but it's only
+        called from contexts where the lock is already held.
+        """
+        self.telemetry["size"] = len(self._cache)
+
+
+class SFDictFileCache(SFDictCache):
+    def __init__(
+        self,
+        file_path: str | dict[str, str],
+        entry_lifetime: int = constants.DAY_IN_SECONDS,
+        file_timeout: int = 0,
+    ) -> None:
+        """Inits an SFDictFileCache with path, lifetime.
+
+        File path can be a dictionary that contains different paths for different OSes,
+        possible keys are: 'darwin', 'linux' and 'windows'. If a current platform
+        cannot be determined, or is not in the dictionary we'll use the first value.
+
+        Once we select a location based on file path, we write and read a random
+        temporary file to check for read/write permissions. If this fails OSError might
+        be thrown.
+        """
+        super().__init__(
+            entry_lifetime=entry_lifetime,
+        )
+        if isinstance(file_path, str):
+            self.file_path = os.path.expanduser(file_path)
+        else:
+            current_platform = platform.system().lower()
+            if current_platform is None or current_platform not in file_path:
+                self.file_path = next(iter(file_path.values()))
+            else:
+                self.file_path = os.path.expanduser(file_path[current_platform])
+        # Once we decided on where to put the file cache make sure that this
+        #  place is readable/writable by us
+        random_string = "".join(random.choice(string.ascii_letters) for _ in range(5))
+        tmp_file, tmp_file_path = tempfile.mkstemp(
+            dir=os.path.dirname(self.file_path),
+        )
+        try:
+            with open(tmp_file, "w") as w_file:
+                # Might raise an OSError
+                w_file.write(random_string)
```

Comment:
Do we need to wrap the error and give a clearer error message like "Specified file path is not readable/writable"?

Reply:
Sure, I can add this